scenario/context.py

```diff
@@ -600,42 +589,29 @@ def _record_status(self, state: "State", is_app: bool):
         else:
             self.unit_status_history.append(state.unit_status)
 
-    def manager(self, event: "_Event", state: "State"):
+    def __call__(self, event: "_Event", state: "State"):
         """Context manager to introspect live charm object before and after the event is emitted.
 
         Usage::
 
             ctx = Context(MyCharm)
-            with ctx.manager(ctx.on.start(), State()) as manager:
-                assert manager.charm._some_private_attribute == "foo"  # noqa
-                manager.run()  # this will fire the event
-                assert manager.charm._some_private_attribute == "bar"  # noqa
+            with ctx(ctx.on.start(), State()) as event:
+                event.charm._some_private_setup()
+                event.run()  # this will fire the event
+                assert event.charm._some_private_attribute == "bar"  # noqa
+
+            with ctx(Action("foo"), State()) as event:
+                event.charm._some_private_setup()
+                event.run_action()  # this will fire the event
```

Comment:
it does feel a little bit weird that you have to call a different method depending on the type of what you pass to `ctx()`. 

The alternative of having `event.run()` return `State|ActionOutput` is however worse. 

What happens if you call the wrong one?

Reply:
I see the `use run() to run non-action events` error, but the reverse is still uncovered.

Reply:
> it does feel a little bit weird that you have to call a different method depending on the type of what you pass to `ctx()`.

I feel it's better that the different name is there when you're emitting the event than having it when you're creating the context - it feels more similar to the `ctx.run` and `ctx.run_action` pair.

> The alternative of having `event.run()` return `State|ActionOutput` is however worse.

Agreed.

The alternative would be a way bigger change where the action output goes somewhere else and you always get a `State` from `run` and `run_action` (in a context manager or not). The action logs/results clearly don't belong in the `State`, so presumably the only place they could go would be into some new attribute of the `Context` like the history ones, Juju logs, versions, etc (`Context.action_result_history`, `Context.action_log_history` or something). An action test would then look like:

```python
ctx = Context(MyCharm)

state = ctx.run(ctx.on.action("backup_db", params={...}), State())

assert state...
assert ctx.action_results_history == [{"filename": "/tmp/backup.tar.gz"}]
assert ctx.action_log_history == ["log 1", "log 2"]
```

I don't hate the logs (it feels natural enough that they're alongside juju-log ones). I don't particularly like having the results there.

> I see the `use run() to run non-action events` error, but the reverse is still uncovered.

Ah, I think I've put both there for the normal `Context.run` and `Context.run_action` but both are missing for the manager cases. Will fix.

Reply:
with juju cli, unless you force `--wait`, the action output is also cached and can be accessed by doing `juju show-operation [action ID]` 

we could store the results in `ctx._action_results`, the logs in `ctx._action_logs` and make them available at `ctx.get_operation(Action)` to have sort of a similar API.

user code:

```python
state = ctx.run(ctx.on.install(), State())
op = ctx.get_operation()  # ERROR since we didn't just run an action


state = ctx.run(ctx.on.action("backup_db", params={...}), State())
op = ctx.get_operation()
assert op.logs == ...
assert op.return_code == 0
assert op.results == {...}


# or, without the intermediate object:
state = ctx.run(ctx.on.action("backup_db", params={...}), State())
assert ctx.last_operation.logs == ...
assert ctx.last_operation.return_code == 0
assert ctx.last_operation.results == {...}
```

Reply:
I really like the idea of unifying run and run_action, and @benhoyt does as well.

Rather than a `get_operation` method, I've proposed having a `Task` class that wraps the results and logs much like `ActionOutput` did, and having a history of those in the context. @benhoyt and I talked about this before I did any implementing and the initial thought was to just have the latest one, but I think I like having a list instead - it feels closer to what Juju actually does.

If people like `Operation` more than `Task` as a class name, I'm ok to switch. Juju has an operation that has a list of tasks - I don't think we want to model it that closely, so just picking one seems best. It feels to me that in Scenario it's really more the task than the operation (e.g. it's modelling it on a single unit). We could keep `ActionOutput` but renaming also works around an annoyance that when the Scenario objects are in the `ops.testing` namespace there's a clash with the existing Harness `ActionOutput`.